src/shuffler-codereview.c

```diff
@@ -0,0 +1,1228 @@
+    u16 objNum = gSpecialVar_0x8008 - 1;
+    u16 flagId = AdjustedTemplates[objNum].flagId;
+    if (FlagClear(flagId) != 0 || FlagGet(flagId) != 0) {
+        MYLOG("NPC Flag wasn't cleared properly! %X", flagId);
+    }
+    return 0;
+}
+
+u16 DoWitchDeal(void) {
+    u16 objNum = gSpecialVar_0x8008 - 1;
+
+    bool8 broke = FALSE;
+    int status = STATUS1_POISON;
+    for (int i = 0; i < gPlayerPartyCount; i++) {
+        if (!GetMonData(&gPlayerParty[i], MON_DATA_IS_EGG) && gPlayerParty[i].hp > 0) {
+            SetMonData(&gPlayerParty[i], MON_DATA_STATUS, &status);
+        }
+    }
+
+    AddBagItem(AdjustedObjects[objNum].npc.witch.itemReward.item, AdjustedObjects[objNum].npc.witch.itemReward.quantity);
+}
+
+u16 BufferWitchText(void) {
+    u16 objNum = gSpecialVar_0x8008 - 1;
+    StringCopy(gStringVar1, AdjustedObjects[objNum].npc.witch.effectText);
+    StringCopy(gStringVar2, AdjustedObjects[objNum].npc.witch.rewardText);
+}
+
+u16 ShuffledFlagNumberByObjectEventId(u16 objNum) {
+    if (RedirectedTrainerFlags & (1 << (objNum - 1))) {
+        return TEMP_FLAGS_START + AdjustedTemplates[objNum - 1].flagId;
+    }
+    return TEMP_FLAGS_START + objNum;
+}
+
+// determines the level scaling of the room based on the distance the player
+// has travelled. this function is likely to be rewritten many times before I
+// settle on anything.
+u8 GetScaledLevel(u8 dist) {
+    switch (dist) {
+    case 0:
+    case 1:
+    case 2:
+    case 3:
+    case 4:
+    case 5:
+        // first set of common rooms
+        return 4;
+    case 6:
+    case 7:
+    case 8:
+    case 9:
+    case 10:
+    case 11:
+        // first gym
+        return 8;
+    case 12:
+    case 13:
+    case 14:
+    case 15:
+    case 16:
+        // second set of common rooms
+        return 10;
+    case 17:
+    case 18:
+    case 19:
+    case 20:
+    case 21:
+    case 22:
+        // second gym
+        return 14;
+    case 23:
+    case 24:
+    case 25:
+    case 26:
+    case 27:
+        // third set of common rooms
+        return 16;
+    case 28:
+    case 29:
+    case 30:
+    case 31:
+    case 32:
+    case 33:
+        // third gym
+        return 20;
+    default:
+        return 30;
+    }
+}
+
+u16 AdjustSpecies(u16 species, u8 level, int evoStrat) {
+    u16 newSpecies = species;
+    int possible = 0;
+    u16 possibilities[EVOS_PER_MON + 1];
+    int addSelf = 0;
+
+    for (int i = 0; i < EVOS_PER_MON; i++) {
+        u16 method = gEvolutionTable[species][i].method;
+        u16 param = gEvolutionTable[species][i].param;
+        u16 targetSpecies = gEvolutionTable[species][i].targetSpecies;
+        if (!method) {
+            continue;
+        }
+        if (method == EVO_MEGA_EVOLUTION || method == EVO_MOVE_MEGA_EVOLUTION || method == EVO_PRIMAL_REVERSION || evoStrat == EVO_STRAT_DO_NOT) {
+            continue;
+        }
+        if ((evoStrat == EVO_STRAT_NO_BELLOSSOM && targetSpecies == SPECIES_BELLOSSOM) ||
```

Comment:
we've had this fight, but I hate these if conditions. They're untestable, and clearly describe some state known to pokemon players that probably has a name.